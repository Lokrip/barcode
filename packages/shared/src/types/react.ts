import { Ref } from "react";

export interface ClassNameType {
    className?: string;
}

export interface WithRef<T> {
    ref?: Ref<T>;
}
