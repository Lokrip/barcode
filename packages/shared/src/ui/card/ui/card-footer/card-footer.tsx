"use client";

import { FC } from "react";

import styles from "./card-footer.module.scss";
import {
    CardFooterProps,
    CardFooterRootGenericProps,
} from "../../model/types/card-footer-props";
import { correctClass } from "../../../../utils";
import clsx from "clsx";
import { CardAvatar } from "../card-avatar";
import { CardActions } from "../card-actions";
import useSlot from "../../../../utils/use-slot";

export const CardFooterRoot: FC<CardFooterRootGenericProps> = ({
    ownerState,
    className,
    ref,
}) => {
    const { actions, avatar } = ownerState || "";

    const slotAvatar = "avatar";
    const slotActions = "actions";

    const Slot = useSlot(
        <>
            <CardAvatar slot={slotAvatar} avatar={avatar} />
            <CardActions slot={slotActions} actions={actions} />
        </>
    );

    const classNameValid = correctClass(
        clsx(styles.footer, "flex-between"),
        className ?? ""
    );

    return (
        <div
            data-testid="card-footer-root"
            className={classNameValid}
            ref={ref}
        >
            <Slot name={slotAvatar} />
            <Slot name={slotActions} />
        </div>
    );
};

export const CardFooter = (props: CardFooterProps) => {
    const { className, ref, ...other } = props;

    const ownerState = { ...props };

    return (
        <CardFooterRoot
            className={className}
            ownerState={ownerState}
            ref={ref}
            {...other}
        />
    );
};
