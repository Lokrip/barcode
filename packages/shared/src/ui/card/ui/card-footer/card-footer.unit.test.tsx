import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import { CardFooter } from "./card-footer";
import { createRef } from "react";

vi.mock("../card-avatar", () => ({
    CardAvatar: ({ avatar }: any) => <div data-testid="avatar">{avatar}</div>,
}));

vi.mock("../card-actions", () => ({
    CardActions: ({ actions }: any) => (
        <div data-testid="actions">{actions}</div>
    ),
}));

describe("CardFooter", () => {
    it("should render without crashing", () => {
        render(<CardFooter />);
        const element = screen.getByTestId("card-footer-root");
        expect(element).toBeInTheDocument();
    });

    it("should render correctly with avatar and actions props", () => {
        render(
            <CardFooter
                avatar={<span>Avatar Content</span>}
                actions={<span>Actions Content</span>}
            />
        );
        expect(screen.getByTestId("avatar")).toHaveTextContent(
            "Avatar Content"
        );
        expect(screen.getByTestId("actions")).toHaveTextContent(
            "Actions Content"
        );
    });

    it("should not throw error when avatar and actions are undefined or null", () => {
        expect(() =>
            render(<CardFooter avatar={null} actions={undefined} />)
        ).not.toThrow();
    });

    it("should apply external className", () => {
        render(<CardFooter className="custom-class" />);
        const element = screen.getByTestId("card-footer-root");
        expect(element).toHaveClass("custom-class");
    });

    it("should combine external className with base styles", () => {
        render(<CardFooter className="custom-class" />);
        const element = screen.getByTestId("card-footer-root");
        expect(element.className).toMatch(/flex-between/);
    });

    it("should apply default styles when className is not provided", () => {
        render(<CardFooter />);
        const element = screen.getByTestId("card-footer-root");
        expect(element.className).toMatch(/flex-between/);
    });

    it("should construct and pass ownerState with all props", () => {
        render(
            <CardFooter
                avatar={<span data-testid="avatar-prop">avatar</span>}
                actions={<span data-testid="actions-prop">actions</span>}
            />
        );
        expect(screen.getByTestId("avatar-prop")).toBeInTheDocument();
        expect(screen.getByTestId("actions-prop")).toBeInTheDocument();
    });

    it("should forward ref to the root div element", () => {
        const ref = createRef<HTMLDivElement>();
        render(<CardFooter ref={ref} />);
        expect(ref.current).toBeInstanceOf(HTMLDivElement);
    });

    it("should work correctly when ref is not provided", () => {
        expect(() => render(<CardFooter />)).not.toThrow();
    });

    it("should render correctly with no props", () => {
        render(<CardFooter />);
        const element = screen.getByTestId("card-footer-root");
        expect(element).toBeInTheDocument();
    });

    it("should match snapshot", () => {
        const { asFragment } = render(
            <CardFooter
                avatar={<span>Avatar</span>}
                actions={<span>Actions</span>}
                className="test-class"
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
