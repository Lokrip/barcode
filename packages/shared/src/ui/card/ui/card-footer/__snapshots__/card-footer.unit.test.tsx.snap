// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardFooter > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="_footer_4c9a6c flex-between test-class"
    data-testid="card-footer-root"
  >
    <div
      data-testid="avatar"
    >
      <span>
        Avatar
      </span>
    </div>
    <div
      data-testid="actions"
    >
      <span>
        Actions
      </span>
    </div>
  </div>
</DocumentFragment>
`;
